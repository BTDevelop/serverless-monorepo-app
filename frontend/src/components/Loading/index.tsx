import * as React from 'react';
import { LoadingComponentProps } from 'react-loadable';
import { Button, Dimmer, Loader, Segment } from 'semantic-ui-react';

const Loading = () => (
  <Dimmer active={true} inverted={true}>
    <Loader inverted={true}>Loading</Loader>
  </Dimmer>
);

export const ComponentLoading = (props: LoadingComponentProps) => {
  if (props.error) {
    return (
      <Segment>
        Error! <Button onClick={props.retry}>Retry</Button>
      </Segment>
    );
  } else if (props.timedOut) {
    return (
      <Segment>
        Taking a long time... <Button onClick={props.retry}>Retry</Button>
      </Segment>
    );
  } else if (props.pastDelay) {
    return <Segment>Loading...</Segment>;
  } else {
    return null;
  }
};

export default Loading;
