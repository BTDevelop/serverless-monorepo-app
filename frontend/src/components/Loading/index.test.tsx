import { shallow } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Button } from 'semantic-ui-react';
import Loading, { ComponentLoading } from './';

describe('Loading Components', () => {
  describe('Loading', () => {
    test('should render correctly', () => {
      const tree = renderer.create(<Loading />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('ComponentLoading', () => {
    const props = {
      error: null,
      isLoading: false,
      pastDelay: false,
      retry: jest.fn(),
      timedOut: false,
    };

    beforeEach(() => {
      jest.clearAllMocks();
    });

    test('should render correctly', () => {
      const tree = renderer.create(<ComponentLoading {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should render correctly with error', () => {
      const tree = renderer
        .create(<ComponentLoading {...props} error={new Error('error')} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should render correctly with timedOut', () => {
      const tree = renderer
        .create(<ComponentLoading {...props} timedOut={true} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should render correctly with pastDelay', () => {
      const tree = renderer
        .create(<ComponentLoading {...props} pastDelay={true} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should retry on click (error)', () => {
      const wrapper = shallow(
        <ComponentLoading {...props} error={new Error('error')} />,
      );

      wrapper.find(Button).simulate('click');

      expect(props.retry).toHaveBeenCalledTimes(1);
    });

    test('should retry on click (timeout)', () => {
      const wrapper = shallow(<ComponentLoading {...props} timedOut={true} />);

      wrapper.find(Button).simulate('click');

      expect(props.retry).toHaveBeenCalledTimes(1);
    });
  });
});
