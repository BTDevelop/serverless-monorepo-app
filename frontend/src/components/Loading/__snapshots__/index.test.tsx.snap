// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading Components ComponentLoading should render correctly 1`] = `null`;

exports[`Loading Components ComponentLoading should render correctly with error 1`] = `
<div
  className="ui segment"
>
  Error! 
  <button
    className="ui button"
    onClick={[Function]}
    role="button"
  >
    Retry
  </button>
</div>
`;

exports[`Loading Components ComponentLoading should render correctly with pastDelay 1`] = `
<div
  className="ui segment"
>
  Loading...
</div>
`;

exports[`Loading Components ComponentLoading should render correctly with timedOut 1`] = `
<div
  className="ui segment"
>
  Taking a long time... 
  <button
    className="ui button"
    onClick={[Function]}
    role="button"
  >
    Retry
  </button>
</div>
`;

exports[`Loading Components Loading should render correctly 1`] = `
<div
  className="ui active transition visible inverted dimmer"
  onClick={[Function]}
>
  <div
    className="content"
  >
    <div
      className="ui inverted text loader"
    >
      Loading
    </div>
  </div>
</div>
`;
